test/operator/operator_test.go

```diff
@@ -54,3 +54,28 @@ func TestOperator_interface(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, true, output)
 }
+
+func TestOperator_Function(t *testing.T) {
+	env := map[string]interface{}{
+		"foo": Value{1},
+		"bar": Value{2},
+	}
+
+	program, err := expr.Compile(
+		`foo + bar`,
+		expr.Env(env),
+		expr.Operator("+", "Add"),
+		expr.Function("Add", func(args ...interface{}) (interface{}, error) {
```

Comment:
Great. I just started experimenting with the lib, so I  didn't notice a lack of type definitions in the test. I’ll fix the pull request accordingly.

Reply:
I've added the proposed type definition but needed help figuring out how to benefit from it. Because function type information is stored in the `Functions` map, and I used data from `Types` in all the places I edited in this pull request.
I could wrap variadic function from `Function` definition into the concrete provided type. Still, I did not figure out what to do if several types are provided because the constructed type variants would collide by the name in the `Type` map.

Reply:
It is the tricky part. Functions can come from two places: TypesTable and Function. 

So we need to check both with the same preferences as the code execution (BTW will be cool to add explicit test for func precedence as well).